torchao/prototype/awq/layout.py

```diff
@@ -0,0 +1,151 @@
+import torch
+from typing import Callable, Optional, Dict, Any
+from torch.utils._python_dispatch import return_and_correct_aliasing
+from torchao.utils import (
+    TorchAOBaseTensor,
+    TORCH_VERSION_AT_LEAST_2_5,
+)
+from torchao.dtypes import AffineQuantizedTensor
+
+
+aten = torch.ops.aten
+
+
+class WeightTensorWithEqualizationScales(TorchAOBaseTensor):
+    """
+    Tensor subclass that wraps a quantized weight tensor and provides the equalization scales which are applied to activations.
+
+    Args:
+        quantized_weight_tensor (torch.Tensor): The weight tensor to be wrapped.
+        scale (torch.Tensor): The scale tensor for activation quantization.
+        zero_point (Optional[torch.Tensor]): The zero point tensor for activation quantization. Default is None.
+        equalization_scale (torch.Tensor): The equalization scale tensor.
+    """
+
+    quantized_weight_tensor: AffineQuantizedTensor
+    equalization_scale: torch.Tensor
+
+    def __new__(
+        cls,
+        quantized_weight_tensor: torch.Tensor,
+        equalization_scale: torch.Tensor
+    ):
+        kwargs = {}
+        dtype = quantized_weight_tensor.dtype
+        kwargs["dtype"] = dtype
+        kwargs["requires_grad"] = False
+        kwargs["device"] = quantized_weight_tensor.device
+        shape = quantized_weight_tensor.shape
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        quantized_weight_tensor: torch.Tensor,
+        equalization_scale: torch.Tensor
+    ):
+        self.quantized_weight_tensor = quantized_weight_tensor
+        self.equalization_scale = equalization_scale
+
+    def __repr__(self):
+        return f"LinearActivationQuantizedTensor({self.quantized_weight_tensor}, eq_scale={self.equalization_scale})"
+
+    def __tensor_flatten__(self):
+        tensor_data = ["quantized_weight_tensor", "equalization_scale"]
+        return tensor_data, []
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        quantized_weight_tensor = tensor_data_dict["quantized_weight_tensor"]
+        equalization_scale = tensor_data_dict["equalization_scale"]
+        return cls(
+            quantized_weight_tensor,
+            equalization_scale,
+        )
+
+    @staticmethod
+    def _quantized_linear_op(
+        input_tensor: torch.Tensor, weight_tensor: torch.Tensor, bias: torch.Tensor
+    ):
+        return torch.nn.functional.linear(
+            input_tensor / weight_tensor.equalization_scale, weight_tensor.quantized_weight_tensor.dequantize(), bias
```

Comment:
should probably not dequantize here I think, we should be able to reuse the optimized implementation for linear that comes with `weight_tensor.quantized_weight_tensor`, I feel you could just use the class from https://github.com/pytorch/ao/pull/990 though

Reply:
I see yea the dequantize is not needed